File: examples/dekita/.claude/hooks/research_requirement_check.py
#!/usr/bin/env python3
"""Issue/PR作成前にWeb調査を強制。

Why:
    十分な調査なしにIssue/PRを作成すると、既存の解決策を見落としたり、
    同様のIssueが既に存在する可能性がある。事前調査を強制する。

What:
    - gh issue create / gh pr create コマンドを検出
    - セッション内のWebSearch/WebFetch履歴を確認
    - 調査なしの場合はブロック
    - バイパス条件: documentation/trivialラベル、十分な探索、本文に「調査不要」

Remarks:
    - ブロック型フック（PreToolUse:Bash）
    - research-trackerがWeb調査を記録、本フックが検証
    - コードベース探索（Grep/Glob 5回以上）もWeb調査の代替として許可

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#1957: ラベル抽出をlib/labels.pyに共通化
    - silenvx/dekita#2578: heredocパターンを先にチェック（ネストされた引用符対応）
"""

import json
import re
import sys
from pathlib import Path

# Add hooks directory to path for common imports
sys.path.insert(0, str(Path(__file__).parent))

from common import check_research_done, get_exploration_depth
from lib.execution import log_hook_execution
from lib.labels import extract_labels_from_command, split_comma_separated_labels
from lib.results import make_approve_result, make_block_result
from lib.session import parse_hook_input
from lib.strings import strip_quoted_strings

HOOK_NAME = "research-requirement-check"

# Labels that bypass research requirement (simple tracking issues, chores, docs)
BYPASS_LABELS = {"documentation", "trivial"}

# Keywords in body that bypass research requirement
BYPASS_KEYWORDS = {"調査不要", "no research needed", "skip research", "調査済み"}


def is_gh_issue_create(command: str) -> bool:
    """Check if command is gh issue create."""
    stripped = strip_quoted_strings(command)
    return bool(re.search(r"\bgh\s+issue\s+create\b", stripped))


def is_gh_pr_create(command: str) -> bool:
    """Check if command is gh pr create."""
    stripped = strip_quoted_strings(command)
    return bool(re.search(r"\bgh\s+pr\s+create\b", stripped))


def extract_labels(command: str) -> set[str]:
    """Extract --label values from command.

    Handles:
    - Multiple --label flags: --label bug --label urgent
    - Comma-separated labels: --label "tracking,backend"
    - Comma-separated with spaces: --label "tracking, backend"
    - Single labels: --label tracking

    Issue #1957: Use shared shlex-based implementation from lib/labels.py.
    """
    # Use shared functions for robust label extraction
    raw_labels = extract_labels_from_command(command)
    split_labels = split_comma_separated_labels(raw_labels)
    # Convert to lowercase set (original behavior)
    return {label.lower() for label in split_labels}


def extract_body(command: str) -> str:
    """Extract --body value from command.

    Handles:
    - Heredoc style: --body "$(cat <<'EOF' ... EOF)" (checked first to handle nested quotes)
    - Double-quoted: --body "message" or -b "message"
    - Single-quoted: --body 'message' or -b 'message'
    - Escaped quotes within strings

    Issue #2578: Heredoc pattern must be checked FIRST because:
    - Heredoc content may contain nested quotes (e.g., code examples with --body "...")
    - Simple quote pattern would incorrectly match at the first nested quote
    - This caused "調査不要" keyword to not be detected when it appears after nested quotes
    """
    # Check for heredoc style FIRST: --body "$(cat <<'EOF' ... EOF)"
    # This handles nested quotes within heredoc content correctly
    match = re.search(r"--body\s+\"?\$\(cat\s+<<['\"]?EOF['\"]?\s*([\s\S]*?)EOF", command)
    if match:
        return match.group(1)
    # Match --body "..." (handles escaped quotes with \\")
    match = re.search(r'(?:--body|-b)\s+"((?:[^"\\]|\\.)*)"', command)
    if match:
        return match.group(1).replace('\\"', '"')
    # Match --body '...' (handles escaped quotes with \\')
    match = re.search(r"(?:--body|-b)\s+'((?:[^'\\]|\\.)*)'", command)
    if match:
        return match.group(1).replace("\\'", "'")
    return ""


def has_bypass_label(command: str) -> bool:
    """Check if command has a bypass label."""
    labels = extract_labels(command)
    return bool(labels & BYPASS_LABELS)


def has_bypass_keyword(command: str) -> bool:
    """Check if command body contains bypass keyword."""
    body = extract_body(command).lower()
    return any(keyword.lower() in body for keyword in BYPASS_KEYWORDS)


def main() -> None:
    """Main entry point for the hook."""
    try:
        input_data = parse_hook_input()
    except json.JSONDecodeError:
        # Invalid input - approve silently
        print(json.dumps({"decision": "approve"}))
        return

    tool_name = input_data.get("tool_name", "")

    # Only check Bash commands
    if tool_name != "Bash":
        print(json.dumps({"decision": "approve"}))
        return

    command = input_data.get("tool_input", {}).get("command", "")

    # Check if this is gh issue create or gh pr create
    is_issue = is_gh_issue_create(command)
    is_pr = is_gh_pr_create(command)

    if not (is_issue or is_pr):
        print(json.dumps({"decision": "approve"}))
        return

    action_type = "Issue" if is_issue else "PR"

    # Check bypass conditions
    if has_bypass_label(command):
        log_hook_execution(
            HOOK_NAME,
            "approve",
            reason=f"{action_type}作成を許可（バイパスラベル）",
            details={"command_preview": command[:80]},
        )
        result = make_approve_result(HOOK_NAME, f"{action_type}作成を許可（バイパスラベル）")
        print(json.dumps(result))
        return

    if has_bypass_keyword(command):
        log_hook_execution(
            HOOK_NAME,
            "approve",
            reason=f"{action_type}作成を許可（バイパスキーワード）",
            details={"command_preview": command[:80]},
        )
        result = make_approve_result(HOOK_NAME, f"{action_type}作成を許可（調査不要）")
        print(json.dumps(result))
        return

    # Check if research was done
    if check_research_done():
        log_hook_execution(
            HOOK_NAME,
            "approve",
            reason=f"{action_type}作成を許可（調査済み）",
            details={"command_preview": command[:80]},
        )
        result = make_approve_result(HOOK_NAME, f"{action_type}作成を許可（調査済み）")
        print(json.dumps(result))
        return

    # Check if exploration is sufficient (alternative to web research)
    exploration = get_exploration_depth()
    if exploration.get("sufficient", False):
        log_hook_execution(
            HOOK_NAME,
            "approve",
            reason=f"{action_type}作成を許可（十分な探索）",
            details={
                "command_preview": command[:80],
                "exploration_total": exploration.get("total", 0),
            },
        )
        result = make_approve_result(
            HOOK_NAME,
            f"{action_type}作成を許可（探索{exploration.get('total', 0)}回）",
        )
        print(json.dumps(result))
        return

    # Block: No research or exploration done
    log_hook_execution(
        HOOK_NAME,
        "block",
        reason=f"{action_type}作成をブロック（調査なし）",
        details={
            "command_preview": command[:80],
            "exploration_total": exploration.get("total", 0),
        },
    )

    block_message = f"""Web検索/情報収集が行われていません。

{action_type}作成前に、以下のツールで調査してください:

**推奨ツール:**
- Context7: 最新ライブラリドキュメントの取得
  → mcp__context7__resolve-library-id, mcp__context7__get-library-docs
- WebSearch: 最新情報・ベストプラクティスの検索
- Grep/Glob: 既存コードベースのパターン調査（現在 {exploration.get("total", 0)}/5回）
- Task(Explore): コードベース構造の理解

**調査不要の場合:**
- --label documentation または --label trivial を付与
- 本文に「調査不要」を含める"""

    result = make_block_result(HOOK_NAME, block_message)
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/research_requirement_check.ts
#!/usr/bin/env bun
/**
 * Issue/PR作成前にWeb調査を強制する。
 *
 * Why:
 *   十分な調査なしにIssue/PRを作成すると、既存の解決策を見落としたり、
 *   同様のIssueが既に存在する可能性がある。事前調査を強制する。
 *
 * What:
 *   - gh issue create / gh pr create コマンドを検出
 *   - セッション内のWebSearch/WebFetch履歴を確認
 *   - 調査なしの場合はブロック
 *   - バイパス条件: documentation/trivialラベル、十分な探索、本文に「調査不要」
 *
 * Remarks:
 *   - ブロック型フック（PreToolUse:Bash）
 *   - research-trackerがWeb調査を記録、本フックが検証
 *   - コードベース探索（Grep/Glob 5回以上）もWeb調査の代替として許可
 *   - Python版: research_requirement_check.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#1957: ラベル抽出をlib/labels.pyに共通化
 *   - silenvx/dekita#2578: heredocパターンを先にチェック
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { formatError } from "../lib/format_error";
import { extractLabelsFromCommand, splitCommaSeparatedLabels } from "../lib/labels";
import { logHookExecution } from "../lib/logging";
import { checkResearchDone, getExplorationDepth, getSessionDir } from "../lib/research";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { createHookContext, parseHookInput } from "../lib/session";
import { stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "research-requirement-check";

// Labels that bypass research requirement
const BYPASS_LABELS = new Set(["documentation", "trivial"]);

// Keywords in body that bypass research requirement
const BYPASS_KEYWORDS = ["no research needed", "skip research", "research complete"];

/**
 * Check if command is gh issue create.
 */
export function isGhIssueCreate(command: string): boolean {
  const stripped = stripQuotedStrings(command);
  return /\bgh\s+issue\s+create\b/.test(stripped);
}

/**
 * Check if command is gh pr create.
 */
export function isGhPrCreate(command: string): boolean {
  const stripped = stripQuotedStrings(command);
  return /\bgh\s+pr\s+create\b/.test(stripped);
}

/**
 * Extract --label values from command.
 */
export function extractLabels(command: string): Set<string> {
  const rawLabels = extractLabelsFromCommand(command);
  const splitLabels = splitCommaSeparatedLabels(rawLabels);
  return new Set(splitLabels.map((l) => l.toLowerCase()));
}

/**
 * Extract --body value from command.
 *
 * Heredoc pattern must be checked FIRST because:
 * - Heredoc content may contain nested quotes
 * - Simple quote pattern would incorrectly match at the first nested quote
 */
export function extractBody(command: string): string {
  // Check for heredoc style FIRST: --body "$(cat <<'DELIMITER' ... DELIMITER)"
  // Use dynamic delimiter capture (group 1) and backreference (\1) to support
  // any delimiter (EOF, END, TEXT, END-OF-FILE, etc.), not just hardcoded 'EOF'
  // Use [ \t]*\r?\n to strictly match horizontal whitespace and required newline
  // (prevents consuming body content indentation with greedy \s*)
  // Use (?:([\s\S]*?)\r?\n)? to handle empty bodies where delimiter immediately follows opening newline
  // Use \1 to ensure delimiter matches at line start (prevents matching delimiter word inside body)
  // Use lookahead (?=\r?\n|\)|$) to ensure delimiter is the only content on the line
  const heredocMatch = command.match(
    /(?:--body|-b)\s+"?\$\(cat\s+<<\s*['"]?([\w-]+)['"]?[ \t]*\r?\n(?:([\s\S]*?)\r?\n)?\1(?=\r?\n|\)|$)/,
  );
  if (heredocMatch) {
    return heredocMatch[2] || "";
  }

  // Match --body "..." (handles escaped quotes)
  const dqMatch = command.match(/(?:--body|-b)\s+"((?:[^"\\]|\\.)*)"/);
  if (dqMatch) {
    return dqMatch[1].replace(/\\"/g, '"');
  }

  // Match --body '...'
  const sqMatch = command.match(/(?:--body|-b)\s+'((?:[^'\\]|\\.)*)'/);
  if (sqMatch) {
    return sqMatch[1].replace(/\\'/g, "'");
  }

  return "";
}

/**
 * Check if command has a bypass label.
 */
export function hasBypassLabel(command: string): boolean {
  const labels = extractLabels(command);
  for (const label of labels) {
    if (BYPASS_LABELS.has(label)) {
      return true;
    }
  }
  return false;
}

/**
 * Check if command body contains bypass keyword.
 */
export function hasBypassKeyword(command: string): boolean {
  const body = extractBody(command).toLowerCase();
  return BYPASS_KEYWORDS.some((keyword) => body.includes(keyword.toLowerCase()));
}

async function main(): Promise<void> {
  let sessionId: string | undefined;

  try {
    const data = await parseHookInput();
    const ctx = createHookContext(data);
    sessionId = ctx.sessionId;
    const toolName = (data.tool_name as string) || "";

    // Only check Bash commands
    if (toolName !== "Bash") {
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Check if this is gh issue create or gh pr create
    const isIssue = isGhIssueCreate(command);
    const isPr = isGhPrCreate(command);

    if (!isIssue && !isPr) {
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const actionType = isIssue ? "Issue" : "PR";

    // Check bypass conditions
    if (hasBypassLabel(command)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${actionType} creation approved (bypass label)`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME, `${actionType} creation approved (bypass label)`);
      console.log(JSON.stringify(result));
      return;
    }

    if (hasBypassKeyword(command)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${actionType} creation approved (bypass keyword)`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(
        HOOK_NAME,
        `${actionType} creation approved (no research needed)`,
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Get session info for research check
    const sessionIdForResearch = sessionId || undefined;
    const sessionDir = getSessionDir();

    // Check if research was done
    if (checkResearchDone(sessionDir, sessionIdForResearch)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${actionType} creation approved (research done)`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(
        HOOK_NAME,
        `${actionType} creation approved (research done)`,
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Check if exploration is sufficient (alternative to web research)
    const exploration = getExplorationDepth(sessionDir, sessionIdForResearch);
    if (exploration.sufficient) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${actionType} creation approved (sufficient exploration)`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(
        HOOK_NAME,
        `${actionType} creation approved (${exploration.total} explorations)`,
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Block: No research or exploration done
    await logHookExecution(
      HOOK_NAME,
      "block",
      `${actionType} creation blocked (no research)`,
      undefined,
      { sessionId },
    );

    const blockMessage = `No web search/research has been performed.

Before creating ${actionType}, please investigate using the following tools:

**Recommended tools:**
- Context7: Get latest library documentation
  → mcp__context7__resolve-library-id, mcp__context7__get-library-docs
- WebSearch: Search for latest information and best practices
- Grep/Glob: Investigate existing codebase patterns (currently ${exploration.total}/5 times)
- Task(Explore): Understand codebase structure

**If no research is needed:**
- Add --label documentation or --label trivial
- Include "no research needed" in the body`;

    const result = makeBlockResult(HOOK_NAME, blockMessage);
    console.log(JSON.stringify(result));
    process.exit(2);
  } catch (error) {
    // Invalid input or error - approve silently
    console.error(`[${HOOK_NAME}] Error: ${formatError(error)}`);
    const result = makeApproveResult(HOOK_NAME);
    console.log(JSON.stringify(result));
  }
}

if (import.meta.main) {
  main().catch((e) => {
    console.error(`[${HOOK_NAME}] Fatal error: ${formatError(e)}`);
    console.log(JSON.stringify({}));
  });
}
